[registry] Addons styles loader (#6630)
plone/volto: packages/registry/docs/conceptual-guides/add-on-loader.md

---
myst:
  html_meta:
    "description": "An explanation of the add-ons configuration loader in @plone/registry"
    "property=og:description": "An explanation of the add-ons configuration loader in @plone/registry"
    "property=og:title": "Add-ons configuration loader"
    "keywords": "@plone/registry, registry, add-ons, loader"
---

# Add-ons configuration loader

The `@plone/registry` compatible add-ons are able to load configuration into the configuration registry.
This is achieved by using the `main` key entry point specified in the add-on {file}`package.json`.
This should be a JavaScript or TypeScript file, eg. {file}`index.ts` or {file}`index.js` placed somewhere (generally in the root) of your add-on.

```json
{
  "main": "index.ts",
}
```

This file should contain a default export with a function with this signature:

```ts
import type { ConfigType } from '@plone/registry';

export default function loadConfig(config: ConfigType) {
  // You can mutate the configuration object in here
  return config;
}
```

> `@plone/registry` has a mean to generate an add-ons loader file that contains the code needed to load the add-ons configuration of all the registered add-ons, keeping the order in which they were defined.

Reviewer: ```suggestion
`@plone/registry` will generate an add-ons loader file that contains the code needed to load the add-ons configuration of all the registered add-ons, keeping the order in which they were defined.
```

This loader is a JavaScript file and is placed in the root of your application.

> By default, it's called {file}`.registry.loader.js`.

Reviewer: I would avoid starting it with a dot. It makes it a hidden file, so people will miss it.

Author: A bit of history, this used to be a hidden file in the /tmp folder so people can't modify it. It's like this in current Volto. It's generated and managed by `@plone/registry`. Since Vite forces you to have this addressable in the app repo boundaries, I opted by use dots to mark it as "special". I did the same for the new style loader. Does this make sense?

---
myst:
  html_meta:
    "description": "An explanation of the add-ons configuration loader in @plone/registry"
    "property=og:description": "An explanation of the add-ons configuration loader in @plone/registry"
    "property=og:title": "Add-ons configuration loader"
    "keywords": "@plone/registry, registry, add-ons, loader"
---

# Add-ons configuration loader

> The `@plone/registry` compatible add-ons are able to load configuration into the configuration registry.

Reviewer 2: ```suggestion
Add-ons that are compatible with `@plone/registry` can load configuration into the configuration registry.
```

> This is achieved by using the `main` key entry point specified in the add-on {file}`package.json`.

Reviewer 2: ```suggestion
The value of the `main` key entry point in the add-on {file}`package.json` loads configuration.
```

Author: not really, this could be confusing.
So, @plone/registry reads this value to know the source of the loader, not to load anything.

Reviewer 2: Suggestion updated. One more review?

> This should be a JavaScript or TypeScript file, eg. {file}`index.ts` or {file}`index.js` placed somewhere (generally in the root) of your add-on.

Reviewer 2: ```suggestion
This should be a JavaScript or TypeScript file, such as {file}`index.ts` or {file}`index.js`, placed somewhere in your add-on, conventionally at its root.
```

```json
{
  "main": "index.ts",
}
```

This file should contain a default export with a function with this signature:

```ts
import type { ConfigType } from '@plone/registry';

export default function loadConfig(config: ConfigType) {
  // You can mutate the configuration object in here
  return config;
}
```

> `@plone/registry` has a helper utility `createAddonsLoader` for generating an add-ons loader file that contains the code needed to load the add-ons configuration of all the registered add-ons, keeping the order in which they were defined.

Reviewer 2: Let's try to keep sentences short. When it gets too long or runs on, it's probably better to break it into multiple shorter sentences.

```suggestion
`@plone/registry` has a helper utility `createAddonsLoader` which generates an add-ons loader file.
That file contains the code needed to load the add-ons configuration of all the registered add-ons, keeping the order in which they were defined.
```

Author: " which generates an add-ons loader file." -> " which generates **the** add-ons loader file."
I think this is better, the loader file is unique. What do you think?

This loader is a JavaScript file and it is placed in the root of your application.
By default, it's called {file}`.registry.loader.js`.

```{important}
This file is generated and maintained by `@plone/registry`.

> You should not modify it and add your own styles in here.

Reviewer 2: ```suggestion
You should neither modify it nor add your own styles in here.
```

> It will be overwriten in the next bundler run.

Reviewer 2: ```suggestion
It will be overwritten in the next bundler run.
```

```

The add-ons loader generator is meant to be run before bundling your app or by the bundler itself when it runs.

> The `@plone/registry` Vite plugin generates this file so the framework can load it on app bootstrap time.

Reviewer 2: ```suggestion
The `@plone/registry` Vite plugin generates this file, so the framework can load it during app bootstrap time, as shown below.
```

```js
  const projectRootPath = path.resolve('.');
  const { registry, shadowAliases } = AddonRegistry.init(projectRootPath);

  createAddonsLoader(
    registry.getAddonDependencies(),
    registry.getAddons(),
    { tempInProject: true },
  );
```

This will create {file}`.registry.loader.js` in the root of your app.

> Afterwards, you have to make sure that your app loads the code on the app bootstrap, as early as possible in your client (and server) code and as a module side-effect.

Reviewer 2: This sentence should be a new paragraph, so it doesn't merge with the previous code example and sentence that refers to it.

```suggestion

Afterwards, configure your app to load the code during the app bootstrap, as early as possible in both your client and server code, and as a module side-effect, as shown in the following example.
```

```js
import config from '@plone/registry';
import applyAddonConfiguration from './.registry.loader';

applyAddonConfiguration(config);
```

```{note}

> If you are using a Vite-powered framework, just use the `@plone/registry` Vite plugin.

Reviewer 2: ```suggestion
If you use a Vite-powered framework, you can use the `@plone/registry` Vite plugin as a convenience.
```

Author: Well, you should use it yes or yes. I don't know if we could convey this better.

Reviewer 2: I removed weasel words from the suggestion.

> Only in the case that you are using a non-Vite framework, you will have to build your own integration.

Reviewer 2: ```suggestion
If you use a non-Vite framework, you will have to build your own integration.
```

You can take the implementation of the Vite plugin as reference.
```

## Provide optional add-on configurations

You can export additional configuration functions from your add-ons configuration loader file.

> The default export is always loaded by `@plone/registry`, while the named exports are optional for loading, if required.

Reviewer 2: ```suggestion
The default export is always loaded by `@plone/registry`, while the named exports are optional for loading as needed.
```

> The way you specify if these optional loaders must be loaded is through the add-on registration using comma-separated values following a colon after the add-on name:

Reviewer 2: ```suggestion
To specify optional loaders as needed, in the add-on registration, use the name of your add-on, followed by a colon (`:`), followed by the names of the optional loaders as comma-separated values as shown in the following example.
```

```{code-block} json
:emphasize-lines: 4

{
  "name": "my-nice-volto-project",
  "addons": [

>     "volto-foo-add-on:loadOptionalBlocks,overrideSomeDefaultBlock",

Reviewer 2: ```suggestion
    "my-volto-add-on-name:loadOptionalBlocks,overrideSomeDefaultBlock",
```
